server/src/main/java/io/crate/execution/engine/fetch/ReaderBuckets.java

```diff
@@ -102,7 +123,11 @@ private void require(long fetchId) {
 
             @Override
             public boolean hasNext() {
-                return idx < rows.size();
+                boolean hasNext = idx < rows.size();
+                if (!hasNext) {
+                    release();
+                }
```

Comment:
This may be a bit hacky to do this as a side effect.

Could also introduce a ClosableIterator interface and have the BatchIterator use it?

Reply:
> Could also introduce a ClosableIterator interface and have the BatchIterator use it?

Yes I think that sounds like the cleaner option. Any reason why you didn't go in that direction in the first place? Backports?